src/lamplib/src/genny/curator.py

```diff
@@ -145,45 +146,82 @@ def poplar_grpc(cleanup_metrics: bool, workspace_root: str, genny_repo_root: str
+            poplar = subprocess.Popen(
+                args, stdout=log_file_obj, stderr=subprocess.STDOUT, text=True
+            )
+            if poplar.poll() is not None:
+                raise OSError("Failed to start Poplar.")
             try:
-                poplar.terminate()
-                exit_code = poplar.wait(timeout=10)
-                if exit_code not in (0, -15):  # Termination or exit.
-                    raise OSError(f"Poplar exited with code: {exit_code}.")
-            except:
-                # If Poplar doesn't die then future runs can be broken.
-                if poplar.poll() is None:
-                    poplar.kill()
-                raise
+                os.chdir(prior_cwd)
+                connected = False
+                for i in range(10):
+                    time.sleep(0.2)  # sleep to let curator get started. This is a heuristic.
+                    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                    try:
+                        sock.connect(("127.0.0.1", 2288))
+                        # If we didn't throw an exception, then we successfully connected
+                        connected = True
+                        break
+                    except socket.error as socket_error:
+                        if socket_error.errno != errno.ECONNREFUSED:
+                            # We expect to get connection refused if poplar is not up yet.
+                            # If we get something else, bail
+                            raise socket_error
+                    finally:
+                        sock.close()
+                if not connected:
+                    raise OSError(f"Poplar not listening on port 2288")
+                yield poplar
+            finally:
+                try:
+                    poplar.terminate()
+                    exit_code = poplar.wait(timeout=10)
+                    if exit_code not in (0, -15):  # Termination or exit.
+                        raise OSError(f"Poplar exited with code: {exit_code}.")
+                except:
+                    # If Poplar doesn't die then future runs can be broken.
+                    if poplar.poll() is None:
+                        poplar.kill()
+                    _report_poplar_error(log_file_obj)
+                    raise
     finally:
         os.chdir(prior_cwd)
 
 
+def _report_poplar_error(log: TextIO):
+    """Review the log file for fatal errors and print them to stderr.
+
+    We don't want to print the entire log, just highlight the fatal error(s).
+
+    """
+
+    log.seek(0)
+
+    is_oom = False
+
+    for line in log:
+        line = line.strip().lower()
+        if "fatal error" in line:
+            # Error messages should go to sys.stderr. Especially, from the
+            # perspective of DSI, all Genny SLOG log entries are just the stdout of
+            # the child process regardless the log level.
+            print(f"Poplar fatal error: {line}", file=sys.stderr)
```

Comment:
As this is written this seems like it's going to print something out twice, just-in-case? This seems sub-optimal. If SLOG is redirected to a log file, it seems to me that the correct thing to do here is to print this error into the error log?

Additionally, can't we create a bound copy of SLOG that will print out on `stderr`? I'd rather not use `print` logging if we can avoid, I don't want two styles of logging flying around.

Is this change needed for this PR to function?

Reply:
The rationale behind printing error messages to stderr is to enable DSI to classify such messages as errors ( in https://github.com/10gen/dsi/pull/1575).

While is It is feasible to eliminate redundancy and maintain a consistent logging style, I chose not to for two reasons: a) I wanted to keep the changes to a minimum. b) in the future, if we decide to redirect logs to log files and reserve stderr/stdout solely for user interaction (maybe via [DEVPROD-4672](https://jira.mongodb.org/browse/DEVPROD-4672)),   this will no longer pose a problem.

Having said that, I acknowledge the validity of your request and hence have made the necessary updates to the code in the latest commit. [Here](https://parsley.mongodb.com/evergreen/sys_perf_perf_1_node_replSet_longRunning.arm.aws.2023_11_llt_mixed_patch_798243406d1f483e46669d7d20f29f4d8f50ba01_65dfd77c0305b9f4950e16b8_24_02_29_01_05_23/3/task?bookmarks=0,1697&selectedLineRange=L1475-L1518&shareLine=1475) is an example log when run in DSI.